pkg/controller/utils/datadog/datadog.go

```diff
@@ -0,0 +1,441 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+package datadog
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	datadoghqv1alpha1 "github.com/DataDog/datadog-operator/pkg/apis/datadoghq/v1alpha1"
+	"github.com/DataDog/datadog-operator/pkg/controller/utils"
+	api "github.com/zorkian/go-datadog-api"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+const (
+	defaultSendMetricsInterval  = 15 * time.Second
+	defaultMetricsNamespace     = "datadog.operator"
+	gaugeType                   = "gauge"
+	deploymentSuccessValue      = 1.0
+	deploymentFailureValue      = 0.0
+	deploymentMetricFormat      = "%s.%s.deploymentsuccess"
+	stateTagFormat              = "state:%s"
+	clusterNameTagFormat        = "cluster_name:%s"
+	agentName                   = "agent"
+	agentEventName              = "Agent"
+	clusteragentName            = "clusteragent"
+	clusteragentEventName       = "Cluster Agent"
+	clustercheckrunnerName      = "clustercheckrunner"
+	clustercheckrunnerEventName = "Cluster Check Runner"
+	deploymentCreatedEventType  = "created"
+	deploymentDeletedEventType  = "deleted"
+)
+
+var (
+	// ErrEmptyAPIKey empty APIKey error
+	ErrEmptyAPIKey = errors.New("empty api key")
+	// ErrEmptyAPPKey empty APPKey error
+	ErrEmptyAPPKey = errors.New("empty app key")
+)
+
+var log = logf.Log.WithName("DatadogMetricsForwarder")
+
+// MetricsForwarder sends metrics directly to Datadog using the public API
+type MetricsForwarder struct {
+	datadogClient       api.Client
+	k8sClient           client.Client
+	keysHash            uint64
+	retryInterval       time.Duration
+	sendMetricsInterval time.Duration
+	metricsPrefix       string
+	tags                []string
+	componentEnabled    map[string]bool
+	delegator           delegatedAPI
+}
+
+// delegatedAPI is used for testing purpose, it serves for mocking the Datadog API
+type delegatedAPI interface {
+	delegatedSendDeploymentMetric(float64, string, []string) error
```

Comment:
I wonder if maybe it's worth dropping `delegated` on the method names?

Reply:
`MetricsForwarder` already has a method named `sendDeploymentMetric`

Reply:
Ah I see, nm then 👍